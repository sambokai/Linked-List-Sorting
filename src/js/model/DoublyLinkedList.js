class Node {
  constructor(value) {
    this.value = value;
    this.previous = null;
    this.next = null;
  }
}

class DoublyLinkedList {
  constructor(value) {
    this.length = 0;
    this.head = null;
    this.tail = null;

    if (value) { this.append(value); }
  }

  append(value) {
    const node = new Node(value);

    if (!this.head) {
      this.head = node;
      this.tail = node;
      this.length += 1;

      return this;
    }

    node.previous = this.tail;
    this.tail.next = node;
    this.tail = node;
    this.length += 1;

    return this;
  }

  prepend(value) {
    const node = new Node(value);

    if (!this.head) {
      this.head = node;
      this.tail = node;
      this.length += 1;

      return this;
    }

    node.next = this.head;
    this.head.previous = node;

    this.head = node;
    this.length += 1;

    return this;
  }
}

export default DoublyLinkedList;
