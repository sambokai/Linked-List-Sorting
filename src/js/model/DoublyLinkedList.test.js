import DoublyLinkedList from './DoublyLinkedList';

describe('DoublyLinkedList', () => {
  it('should allow creation of an empty doubly linked list', () => {
    const list = new DoublyLinkedList();

    expect(list).toMatchObject({ head: null, length: 0, tail: null });
  });

  describe('append()', () => {
    it('should add a single element to the end of the DoublyLinkedList', () => {
      const list = new DoublyLinkedList();
      const value = 100;

      list.append(value);

      expect(list).toMatchSnapshot();
    });

    it('should add multiple elements to the end of the DoublyLinkedList', () => {
      const list = new DoublyLinkedList();

      list.append(1).append(2).append(3);

      expect(list).toMatchSnapshot();
    });
  });

  describe('prepend()', () => {
    it('should add an element to an empty list', () => {
      const list = new DoublyLinkedList();

      list.prepend(0);

      expect(list).toMatchSnapshot();
    });

    it('should add a single element to the beginning of the DoublyLinkedList', () => {
      const list = new DoublyLinkedList(1);

      list.prepend(0);

      expect(list).toMatchSnapshot();
    });
  });
});
