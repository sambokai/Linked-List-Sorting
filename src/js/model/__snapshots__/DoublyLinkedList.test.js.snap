// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoublyLinkedList append() should add a single element to the end of the DoublyLinkedList 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": null,
    "previous": null,
    "value": 100,
  },
  "length": 1,
  "tail": Node {
    "next": null,
    "previous": null,
    "value": 100,
  },
}
`;

exports[`DoublyLinkedList append() should add multiple elements to the end of the DoublyLinkedList 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": Node {
      "next": Node {
        "next": null,
        "previous": [Circular],
        "value": 3,
      },
      "previous": [Circular],
      "value": 2,
    },
    "previous": null,
    "value": 1,
  },
  "length": 3,
  "tail": Node {
    "next": null,
    "previous": Node {
      "next": [Circular],
      "previous": Node {
        "next": [Circular],
        "previous": null,
        "value": 1,
      },
      "value": 2,
    },
    "value": 3,
  },
}
`;

exports[`DoublyLinkedList prepend() should add a single element to the beginning of the DoublyLinkedList 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": Node {
      "next": null,
      "previous": [Circular],
      "value": 1,
    },
    "previous": null,
    "value": 0,
  },
  "length": 2,
  "tail": Node {
    "next": null,
    "previous": Node {
      "next": [Circular],
      "previous": null,
      "value": 0,
    },
    "value": 1,
  },
}
`;

exports[`DoublyLinkedList prepend() should add an element to an empty list 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": null,
    "previous": null,
    "value": 0,
  },
  "length": 1,
  "tail": Node {
    "next": null,
    "previous": null,
    "value": 0,
  },
}
`;
